lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+	for _, validatorEntryToSkip := range validatorEntriesToSkip {
+		validatorKeysToSkip.Add(string(DBKeyForValidatorByStake(validatorEntryToSkip)))
+	}
+
+	// Retrieve top N active ValidatorEntry PKIDs by stake.
+	key := append([]byte{}, Prefixes.PrefixValidatorByStake...)
+	key = append(key, EncodeUint8(uint8(ValidatorStatusActive))...)
+	_, validatorPKIDsBytes, err := EnumerateKeysForPrefixWithLimitOffsetOrder(
+		handle, key, limit, nil, true, validatorKeysToSkip,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBGetTopActiveValidatorsByStake: problem retrieving top validators: ")
+	}
+
+	// For each PKID, retrieve the ValidatorEntry by PKID.
+	for _, validatorPKIDBytes := range validatorPKIDsBytes {
+		// Convert PKIDBytes to PKID.
+		validatorPKID := &PKID{}
+		exists, err := DecodeFromBytes(validatorPKID, bytes.NewReader(validatorPKIDBytes))
+		if !exists || err != nil {
+			return nil, errors.Wrapf(err, "DBGetTopActiveValidatorsByStake: problem reading ValidatorPKID: ")
+		}
+		// Retrieve ValidatorEntry by PKID.
+		validatorEntry, err := DBGetValidatorByPKID(handle, snap, validatorPKID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "DBGetTopActiveValidatorsByStake: problem retrieving validator by PKID: ")
+		}
+		validatorEntries = append(validatorEntries, validatorEntry)
+	}
+
+	return validatorEntries, nil
+}
+
+func DBGetGlobalStakeAmountNanos(handle *badger.DB, snap *Snapshot) (*uint256.Int, error) {
+	var ret *uint256.Int
+	err := handle.View(func(txn *badger.Txn) error {
+		var innerErr error
+		ret, innerErr = DBGetGlobalStakeAmountNanosWithTxn(txn, snap)
+		return innerErr
+	})
+	return ret, err
+}
+
+func DBGetGlobalStakeAmountNanosWithTxn(txn *badger.Txn, snap *Snapshot) (*uint256.Int, error) {
+	// Retrieve from db.
+	key := DBKeyForGlobalStakeAmountNanos()
+	globalStakeAmountNanosBytes, err := DBGetWithTxn(txn, snap, key)
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return 0.
+		if err == badger.ErrKeyNotFound {
+			return uint256.NewInt(), nil
+		}
+		return nil, errors.Wrapf(err, "DBGetGlobalStakeAmountNanosWithTxn: problem retrieving value")
+	}
+
+	// Decode from bytes.
+	var globalStakeAmountNanos *uint256.Int
+	rr := bytes.NewReader(globalStakeAmountNanosBytes)
+	globalStakeAmountNanos, err = DecodeUint256(rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBGetGlobalStakeAmountNanosWithTxn: problem decoding value")
+	}
+	return globalStakeAmountNanos, nil
+}
+
+func DBPutValidatorWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	validatorEntry *ValidatorEntry,
+	blockHeight uint64,
+) error {
+	if validatorEntry == nil {
+		// This should never happen but is a sanity check.
+		glog.Errorf("DBPutValidatorWithTxn: called with nil ValidatorEntry")
+		return nil
+	}
+
+	// Set ValidatorEntry in PrefixValidatorByPKID.
+	key := DBKeyForValidatorByPKID(validatorEntry)
+	if err := DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, validatorEntry)); err != nil {
```

Comment:
FIXME: I don't think we should be using this block height for anything related to consensus. This is just the blockheight at which the values were flushed, which can vary depending on *when* the flush was called. Right now, if we connect 2 blocks and flush, we'll have a different value than if we connected one at a time, which isn't the case for anything else in our code as of today.

Reply:
Nevermind. I see it's OK because it's just an encoder.